### ISSUE TITLE
Issues with repositories used by creation (likely because org default is not set to `main`)

### ISSUE THREAD

**Issue by mykelalvis** [integrations/terraform-provider-github#899]
This issue regards an organization whose default branch for new repos is changed from GH's default of `main` to `master`

This appears to be due to a change (#562) where `main` is queried rather than checking the org's defaults to see which `ref` is actually created.

### Terraform Version
1.0.5 (also happens with 1.0.6)

### Affected Resource(s)
- `github_branch`
- `github_repository` (when dependent on a branch as a template)

This doesn't appear to be an issue with Terraform core.

### Terraform Configuration Files

```json
{
  "terraform": {
    "required_providers": {
      "github": {
        "source": "integrations/github",
        "version": "~> 4.0"
      }
    }
  }
}
```
Actual version is `4.14.0`

### Debug Output
I am unable to do this for compliance reason.

### Panic Output
It did not

### Expected Behavior

Repo/branch creation and continuation with the rest of the 150+ repository dependency chain

### Actual Behavior
Errors with following: (see factoids below for verification calls using `curl`)

```
│ Error: Error querying GitHub branch reference {ORG}/_default (refs/heads/main): GET https://api.github.com/repos/{ORG}/_default/git/ref/heads/main: 404 Not Found []
│ 
│   with github_branch._default_main,
│   on github_branch.tf.json line 48, in resource.github_branch._default_main:
│   48:     },
│ 

│ Error: POST https://api.github.com/repos/{ORG}/_automation-template/generate: 404 Not Found []
│ 
│   with github_repository.DEPENDENT_REPO,
│   on github_repository.tf.json line 55, in resource.github_repository.DEPENDENT_REPO:
│   55:     },
│ 

```

If it is not clear,`DEPENDENT_REPO` depends on `_default` to complete entirely successfully, and thus the issue is not directly with` github_repository.DEPENDENT_REPO` but rather with the fact that `github_branch._default_main` does not perform a query against the correct API endpoint.

`github_repository` definition for `_default`:

```
    "_default": {
        "auto_init": true,
        "visibility": "private",
        "is_template": true,
        "has_issues": true,
        "has_projects": false,
        "has_wiki": true,
        "name": "_default",
        "delete_branch_on_merge": true
    },
```

`github_branch` for `_default`:

```
    "_default_master": {
        "depends_on": ["github_repository._default"],
        "repository": "_default",
        "branch": "master"
    },
```

`github_repository` definition for `DEPENDENT_REPO`:

```
    "DEPENDENT_REPO": {
        "auto_init": false,
        "template": {
            "owner": "{ORG}",
            "repository": "_default"
        },
        "depends_on": ["github_repository_file.local_file_109271247__default"],
        "visibility": "private",
        "is_template": true,
        "has_issues": true,
        "has_projects": false,
        "has_wiki": true,
        "name": "DEPENDENT_REPO",
        "delete_branch_on_merge": true
    }

```

### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. Create Org
2. Create OAUTH Creds for terraform on org
3. _Set org default branch name to `master` and not `main`_
4. Execute creation of a repo with `auto_init = true`
5. Attempt to set the default branch of that repo to `master` using terraform with `github_branch` and `github_branch_default`
6. Fail

### Important Factoids

This appears to be because a query is being made to `git/ref/heads/main` instead of `git/ref/heads/master`

```
$ curl  -X GET  -H "Authorization: token {TOKEN}"   -H "Accept: application/vnd.github.v3+json"   https://api.github.com/repos/{ORG}/_default/git/refs/heads/master
{
  "ref": "refs/heads/master",
  "node_id": "{NODE_ID}",
  "url": "https://api.github.com/repos/{ORG}/_default/git/refs/heads/master",
  "object": {
    "sha": "{SHA}",
    "type": "commit",
    "url": "https://api.github.com/repos/{ORG}/_default/git/commits/{SHA}"
  }
}
```
vs
```
$ curl  -X GET  -H "Authorization: token {TOKEN}"   -H "Accept: application/vnd.github.v3+json"   https://api.github.com/repos/{ORG}/_default/git/refs/heads/main
{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest"
}
```

Basically, the provider appears to be attempting to query the wrong resource, based on the defaults of the repo at creation time.

### References
- #562 appears to be the underlying cause for this, as the assumed default in that PR is likely not a valid assumption.
- #577 (just because it references the `auto_init` situation in a comment)

**Comment by mykelalvis-rialtic** [comment]
Possibly same/similar issue as #902

**Comment by aidan-mundy** [comment]
@mykelalvis #902 appears to be an entirely different problem. That user is explicitly trying to access a branch labeled `main`.

**Comment by mykelalvis-rialtic** [comment]
Probably so.

**Comment by mykelalvis-rialtic** [comment]
Also, it may be obvious but the workaround for this issue is to refer to initial branch of the [just-created] repo by string value (`"master"`) instead of by reference (`github_branch.blah.branch`) and then load up on `depends_on`.

**Comment by mykelalvis-rialtic** [comment]
Removed spurious issue reference

**Comment by thpang** [comment]
Any updates or progress on this one?
